### PR TITLE
migrate to langgraph

### PR DESCRIPTION
**Pull request by vbarda**

### CODE REVIEW

File: backend/graph.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    chat_history = []
+    for message in messages:
+        if isinstance(message, AIMessage) or isinstance(message, HumanMessage):
+            chat_history.append(message)
+    return chat_history
+
+
+def synthesize_response(
+    state: AgentState, model: LanguageModelLike, prompt_template: str
+):
```
Comment (baskaryan) [inline backend/graph.py:228]: missing return type

File: backend/graph.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    formatted_docs = []
+    for i, doc in enumerate(docs):
+        doc_string = f"<doc id='{i}'>{doc.page_content}</doc>"
+        formatted_docs.append(doc_string)
+    return "\n".join(formatted_docs)
+
+
+def retrieve_documents(state: AgentState):
+    retriever = get_retriever()
+    messages = convert_to_messages(state["messages"])
```
Comment (baskaryan) [inline backend/graph.py:182]: ooc why's this needed? aren't these always BaseMessages

File: backend/graph.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+    relevant_documents = retriever_with_condensed_question.invoke(
+        {"question": query, "chat_history": get_chat_history(messages)}
+    )
+    return {"query": query, "documents": relevant_documents, "messages": []}
+
+
+def route_to_retriever(
+    state: AgentState,
+) -> Literal["retriever", "retriever_with_chat_history"]:
+    if len(state["messages"]) == 1:
```
Comment (baskaryan) [inline backend/graph.py:212]: should this be 0?

File: backend/graph.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    response_synthesizer = prompt | model
+    synthesized_response = response_synthesizer.invoke(
+        {
+            "question": state["query"],
+            "context": format_docs(state["documents"]),
+            "chat_history": get_chat_history(convert_to_messages(state["messages"])),
+        }
+    )
+    return {
+        **state,
```
Comment (baskaryan) [inline backend/graph.py:245]: afaiu nodes only need to return patches to state, not the entire updated state

File: backend/graph.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        "messages": [synthesized_response],
+    }
+
+
+def synthesize_response_default(state: AgentState, config):
```
Comment (baskaryan) [inline backend/graph.py:250]: type hints

File: backend/graph.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        cohere_api_key=os.environ.get("COHERE_API_KEY", "not_provided"),
+    )
+    llm: LanguageModelLike = {
+        OPENAI_MODEL_KEY: gpt_3_5,
+        ANTHROPIC_MODEL_KEY: claude_3_haiku,
+        FIREWORKS_MIXTRAL_MODEL_KEY: fireworks_mixtral,
+        GOOGLE_MODEL_KEY: gemini_pro,
+        COHERE_MODEL_KEY: cohere_command,
+    }[model_name]
+    llm = llm.with_fallbacks(
```
Comment (baskaryan) [inline backend/graph.py:150]: ooc why aren't we using .configurable_alternatives

File: backend/ingest.py [hunk 7]
```diff
@@ -15,6 +13,9 @@
 from langchain_core.embeddings import Embeddings
 from langchain_openai import OpenAIEmbeddings
 
+from backend.constants import WEAVIATE_DOCS_INDEX_NAME
+from backend.parser import langchain_docs_extractor
```
Comment (baskaryan) [inline backend/ingest.py:17]: would be careful about this, vaguely remember that something about the way we deploy makes it so we don't want the `backend.`

**Review by baskaryan — COMMENTED** [review]
this is sweet!

File: backend/graph.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+
+class AgentState(TypedDict):
+    query: str
+    documents: list[Document]
+    messages: Annotated[list[BaseMessage], add_messages]
+
+
+def get_model(model_name: str) -> LanguageModelLike:
+    gpt_3_5 = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0, streaming=True)
```
[the remark this answers is not in the document]
Comment (vbarda) [reply backend/graph.py:118]: wouldn't that be a bit expensive?

File: backend/graph.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+from langchain_fireworks import ChatFireworks
+from langchain_google_genai import ChatGoogleGenerativeAI
+from langchain_openai import ChatOpenAI
+from langgraph.graph import END, StateGraph, add_messages
+
+from backend.constants import WEAVIATE_DOCS_INDEX_NAME
+from backend.ingest import get_embeddings_model
+
+WEAVIATE_URL = os.environ["WEAVIATE_URL"]
+WEAVIATE_API_KEY = os.environ["WEAVIATE_API_KEY"]
```
[the remark this answers is not in the document]
Comment (vbarda) [reply backend/graph.py:32]: agreed, let me update

File: backend/graph.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    if len(state["messages"]) == 1:
+        return "retriever"
+    else:
+        return "retriever_with_chat_history"
+
+
+def get_chat_history(messages: Sequence[BaseMessage]) -> Sequence[BaseMessage]:
+    chat_history = []
+    for message in messages:
+        if isinstance(message, AIMessage) or isinstance(message, HumanMessage):
```
[the remark this answers is not in the document]
Comment (vbarda) [reply backend/graph.py:221]: like a `ToolMessage` in the future

File: backend/graph.py [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    return chat_history
+
+
+def synthesize_response(
+    state: AgentState, model: LanguageModelLike, prompt_template: str
+):
+    prompt = ChatPromptTemplate.from_messages(
+        [
+            ("system", prompt_template),
+            MessagesPlaceholder(variable_name="chat_history"),
```
[the remark this answers is not in the document]
Comment (vbarda) [reply backend/graph.py:232]: nice!

File: backend/graph.py (hunk 1 above)
Comment (vbarda) [reply backend/graph.py:228]: thanks, will fix this and all other missing type annotations

File: backend/graph.py (hunk 2 above)
Comment (vbarda) [reply backend/graph.py:182]: they are coming in as dicts, so this is just deserializing them into `BaseMessage` instances

File: backend/graph.py (hunk 3 above)
Comment (vbarda) [reply backend/graph.py:212]: no, i'll add a note -- at that point in the graph execution there is one message (user input)

File: backend/graph.py (hunk 4 above)
Comment (vbarda) [reply backend/graph.py:245]: ah yes, that's a good point, i'll clean that up

File: backend/graph.py (hunk 6 above)
Comment (vbarda) [reply backend/graph.py:150]: kind of a stylistic preference, i feel like it's a bit verbose to do `.configurable_alternatives` and then an extra `.with_config` just to achieve very simple lookup in this case

File: backend/ingest.py (hunk 7 above)
Comment (vbarda) [reply backend/ingest.py:17]: yea, i think something got messed up in vercel, going to dig deeper

File: backend/graph.py (hunk 6 above)
Comment (baskaryan) [reply backend/graph.py:147]: don't think you'd need the with_config, model selection would happen automatically based on whatever config client passes in. you wouldn't need this get_model(model_name) function at all, you would just create the llm once

File: backend/graph.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (baskaryan) [reply backend/graph.py:220]: wouldn't you want that in chat history? if there's an AIMessage with tool calls the next message has to be ToolMessage or most models will raise an error

File: backend/graph.py (hunk 2 above)
Comment (baskaryan) [reply backend/graph.py:179]: why are they coming in as dicts?

File: backend/graph.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (baskaryan) [reply backend/graph.py:115]: hah good question, not actually sure how expensive 4o is

[the remark this answers is not in the document]
Comment (baskaryan) [reply backend/graph.py:115]: woah still 10x more expensive than 3.5-turbo, good call

File: backend/graph.py (hunk 2 above)
Comment (vbarda) [reply backend/graph.py:179]: I am not actually sure, @nfcampos am i doing something wrong there that it's coming in as dicts or is that an expected behavior?

File: backend/graph.py (hunk 6 above)
Comment (vbarda) [reply backend/graph.py]: that makes perfect sense, thank you -- updated this :)

File: backend/graph.py (hunk 2 above)
Comment (jacoblee93) [reply backend/graph.py:182]: Feels like these should be serialized, yeah

**Comment by jacoblee93** [comment]
Looks good to me if all works!

File: frontend/app/components/ChatWindow.tsx [hunk 12]
```diff
@@ -38,27 +45,92 @@ const MODEL_TYPES = [
...
   const messageContainerRef = useRef<HTMLDivElement | null>(null);
-  const [messages, setMessages] = useState<Array<Message>>([]);
   const [input, setInput] = useState("");
   const [isLoading, setIsLoading] = useState(false);
   const [llm, setLlm] = useState(
     searchParams.get("llm") ?? "openai_gpt_3_5_turbo",
   );
   const [llmIsLoading, setLlmIsLoading] = useState(true);
+  const [assistantId, setAssistantId] = useState<string>("");
+  const [userId, setUserId] = useLocalStorage("userId", null);
```
Comment (vbarda) [inline frontend/app/components/ChatWindow.tsx:74]: @jacoblee93 would love 👀 here and `setUserInfo`

File: frontend/app/components/ChatWindow.tsx [hunk 13]
```diff
@@ -38,27 +45,92 @@ const MODEL_TYPES = [
...
+    try {
+      const assistantId = await getAssistantId(client);
+      setAssistantId(assistantId);
+    } catch (e) {
+      toast.error("Could not load AI agent");
+    }
+  };
+
+  const setUserInfo = () => {
+    if (userId == null) {
```
Comment (jacoblee93) [inline frontend/app/components/ChatWindow.tsx:102]: Any potential race condition with the local storage hook above? I think not but just double checking

Comment (jacoblee93) [reply frontend/app/components/ChatWindow.tsx:102]: Could also use some kind of `getOrCreateUserId` pattern and combine it into your other hook?

Comment (vbarda) [reply frontend/app/components/ChatWindow.tsx:102]: > Any potential race condition with the local storage hook above? I think not but just double checking

don't think so

**Pull request merged**